Modules/_interpretersmodule.c

```diff
@@ -0,0 +1,1892 @@
+            _PyCrossInterpreterData_Release(&item->data);
+            break;
+        }
+        (item + 1)->name = NULL;  // Mark the next one as the last.
+    }
+    if (PyErr_Occurred()) {
+        _sharedns_clear(shared);
+        PyMem_Free(shared);
+        return NULL;
+    }
+    return shared;
+}
+
+static int
+_shareditem_apply(struct _shareditem *item, PyObject *ns)
+{
+    PyObject *name = PyUnicode_FromUnicode(item->name, item->namelen);
+    if (name == NULL) {
+        return 1;
+    }
+    PyObject *value = _PyCrossInterpreterData_NewObject(&item->data);
+    if (value == NULL) {
+        Py_DECREF(name);
+        return 1;
+    }
+    int res = PyDict_SetItem(ns, name, value);
+    Py_DECREF(name);
+    Py_DECREF(value);
+    return res;
+}
+
+// Ultimately we'd like to preserve enough information about the
+// exception and traceback that we could re-constitute (or at least
+// simulate, a la traceback.TracebackException), and even chain, a copy
+// of the exception in the calling interpreter.
+
+struct _shared_exception {
+    char *msg;
+};
+
+static struct _shared_exception *
+_get_shared_exception(void)
+{
+    struct _shared_exception *err = PyMem_NEW(struct _shared_exception, 1);
+    if (err == NULL)
+        return NULL;
+    PyObject *exc;
+    PyObject *value;
+    PyObject *tb;
+    PyErr_Fetch(&exc, &value, &tb);
+    PyObject *msg;
+    if (value == NULL)
+        msg = PyUnicode_FromFormat("%S", exc);
+    else
+        msg = PyUnicode_FromFormat("%S: %S", exc, value);
+    if (msg == NULL) {
+        err->msg = "unable to format exception";
+        return NULL;
+    }
+    err->msg = (char *)PyUnicode_AsUTF8(msg);
+    if (err->msg == NULL) {
+        err->msg = "unable to encode exception";
+    }
+    return err;
+}
+
+static PyObject * RunFailedError;
+
+static int
+interp_exceptions_init(PyObject *ns)
+{
+    // XXX Move the exceptions into per-module memory?
+
+    // An uncaught exception came out of interp_run_string().
+    RunFailedError = PyErr_NewException("_interpreters.RunFailedError",
+                                        PyExc_RuntimeError, NULL);
+    if (RunFailedError == NULL)
+        return -1;
+    PyDict_SetItemString(ns, "RunFailedError", RunFailedError);
+
+    return 0;
+}
+
+static void
+_apply_shared_exception(struct _shared_exception *exc)
+{
+    PyErr_SetString(RunFailedError, exc->msg);
+}
+
+/* channel-specific code */
+
+static PyObject *ChannelError;
+static PyObject *ChannelNotFoundError;
+static PyObject *ChannelClosedError;
+static PyObject *ChannelEmptyError;
+
+static int
+channel_exceptions_init(PyObject *ns)
+{
+    // XXX Move the exceptions into per-module memory?
+
+    // A channel-related operation failed.
+    ChannelError = PyErr_NewException("_interpreters.ChannelError",
+                                      PyExc_RuntimeError, NULL);
+    if (ChannelError == NULL)
+        return -1;
+    PyDict_SetItemString(ns, "ChannelError", ChannelError);
```

Comment:
PyDict_SetItemString can error; check the return code.

Reply:
That's what I get from copy-and-paste from an existing module (see Modules/_threadmodule.c). :/  I'll fix this.